hack/upgrade-test-harness/main.go

```
@@ -0,0 +1,186 @@
+
+	if err := cmd.Execute(); err != nil {
+		zap.S().Errorw("Test failed", "error", err)
+		os.Exit(1)
+	}
+}
+
+func doRun(_ *cobra.Command, _ []string) error {
+	config.InitLogging(opts.logLevel)
+
+	// load test configurations
+	conf, err := config.Load(opts.confFile)
+	if err != nil {
+		return fmt.Errorf("failed to load config: %w", err)
+	}
+
+	// determine the release to test
+	from := mustGetReleasePos(conf, opts.fromRelease)
+	to := mustGetReleasePos(conf, opts.toRelease)
+
+	if to < from {
+		return errInvalidVersionRange
+	}
+
+	// create test context
+	ctx, err := fixture.NewTestContext(kubeConfFlags, int(opts.retryCount), opts.retryDelay, opts.retryTimeout)
+	if err != nil {
+		return err
+	}
+
+	// configure cleanup
+	if !opts.skipCleanup {
+		defer func() {
+			ctx.Info("Starting cleanup")
+
+			if err := ctx.Cleanup(ctx); err != nil {
+				ctx.Errorw("Failed to cleanup", "error", err)
+			}
+		}()
+	}
+
+	// run upgrade fixtures
+	var prevTestParam *fixture.TestParam
+
+	for i := from; i <= to; i++ {
+		currTestParam := conf.TestParam[i]
+
+		// setup upcoming release if necessary
+		if currTestParam.Name == "upcoming" {
+			if err := setupUpcomingRelease(opts.upcomingReleaseYAML); err != nil {
+				return fmt.Errorf("failed to setup upcoming release: %w", err)
+			}
+		}
+
+		fixtures := buildUpgradeFixtures(prevTestParam, currTestParam)
+
+		ctx.Infof("=====[%s]=====", currTestParam.Name)
+
+		for _, f := range fixtures {
+			if err := f.Execute(ctx); err != nil {
+				return err
+			}
+		}
+
+		prevTestParam = &currTestParam
+	}
+
+	// try to scale the last deployed release
+	if prevTestParam != nil {
+		return fixture.TestScaleElasticsearch(*prevTestParam, 5).Execute(ctx)
+	}
+
+	return nil
+}
+
+func mustGetReleasePos(conf *config.File, name string) int {
+	pos, err := conf.ReleasePos(name)
+	if err != nil {
+		panic(err)
+	}
+
+	return pos
+}
+
+func setupUpcomingRelease(installYAML string) error {
+	in, err := os.Open(installYAML)
```

Should this be validated early on to fail fast and not only after running through all the other tests given that `upcoming` will be the last test?